[Doc] Refactored Type system. (#4584)
taichi-dev/taichi: docs/lang/articles/basic/type.md

---
sidebar_position: 3

---

# Type system

Taichi supports two data types in the Taichi scope: primitive type and compound type.

> - Primitive types refer to the various commonly-used numerical data types.

Reviewer: ```suggestion
- Primitive types are referred to as the various commonly-used numerical data types.
```

- Compound types are user-defined data types, which comprise multiple members.

## Primitive types

> Each primitive type is denoted with a character indicating its category followed by a number indicating its precision bits (number of bits for storing the data). The category can be `i` (signed integers), `u` (unsigned integers), or `f` (floating-point numbers); the _precision bits_ can be `8`, `16`, `32`, or `64`. Following are the two most commonly-used types:

Reviewer: ```suggestion
Each primitive type is denoted with a character indicating its category followed by a number indicating its precision bits (number of bits for storing the data). The _category_ can be `i` (signed integers), `u` (unsigned integers), or `f` (floating-point numbers); the _precision bits_ can be `8`, `16`, `32`, or `64`. Following are the two most commonly-used types:
```

- `i32`: 32-bit signed integer
- `f32` : 32-bit floating-point number.

### Supported primitive types

Not all primitive types are supported by your backend. Check out the following table for the supported types. Note that some backends may require extensions to support a specific primitive type.

> | Backend | `i8`               | `i16`              | `i32`              | `i64`              | u8                 | u16                | u32                | u64                | f16                | f32                | f64                |

Reviewer: The table looks good to me. We have to make sure the 'require extension` symbol is consistent across our doc. Currently the operators page still uses the orange one ;/

| ------- | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ |

> | CPU     | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

Reviewer 2: nits:

* heavy_check_mark is dark gray, better use a green color
* `x` and `o` are of the same color, better distinguish the two

Author: Searched for available markdown emojis but failed to find better options. Have asked the frontend developer to come up with CSS for these two emojis.

| Metal   | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :x:                | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :x:                | :x:                | :heavy_check_mark: | :x:                |
| Vulkan  | :o:                | :o:                | :heavy_check_mark: | :o:                | :o:                | :o:                | :heavy_check_mark: | :o:                | :heavy_check_mark: | :heavy_check_mark: | :o:                |

> :o:: Requiring extensions for the backend.

### Default primitive types for integers and floating-point numbers

- The default integer type in Taichi is `ti.i32`.
- The default floating-point type in Taichi is `ti.f32`.

#### Customize the default type

You can change the default primitive types by explicitly specifying the default types when initializing Taichi:

```python
ti.init(default_ip=ti.i64)  # Set the default integer type to ti.i64
ti.init(default_fp=ti.f64)  # Set the default floating-point type to ti.f64
```

#### Set alias

> Taichi supports using`int` as an alias for the default integer type and `float` as an alias for the default floating-point type. In the following example, you change the default primitive types to `i64` and `f64` when initializing Taichi, then you can use `int` to represent `i64` and `float` to represent `f64`.

Reviewer: ```suggestion
Taichi supports using `int` as an alias for the default integer type and `float` as an alias for the default floating-point type. In the following example, you change the default primitive types to `i64` and `f64` when initializing Taichi, then you can use `int` to represent `i64` and `float` to represent `f64`.
```

```python
ti.init(default_ip=ti.i64, default_fp=ti.f64)

x = ti.field(float, 5)
y = ti.field(int, 5)
# is equivalent to:
x = ti.field(ti.f64, 5)
y = ti.field(ti.i64, 5)

def func(a: float) -> int:
    ...
# is equivalent to:
def func(a: ti.f64) -> ti.i64:
    ...
```

### Explicit type casting

> In the Taichi scope, the type of a variable is *static* and *completely determined when it is initialized*. Taichi's compiler does type check at compile time, so you *cannot* change a variable's type once it is initialized. Still, from time to time, you may run into a situation where you have a certain type of data but it is not feasible for an assignment or calculation. Then, you need explicit type casting:

Reviewer: Here i wonder: Should we say 'In Taichi/Python scope' instead of 'in the ..' across the doc? Do we need the article?

Author: I think "in the Taichi scope" is equivalent to "in Taichi" but is more specific. WDYT?  @qiao-bo @k-ye

Yes, we need a definite article here.

Reviewer: In Taichi scope sounds better to me, but i am fine with both options.

> In the Taichi scope, the type of a variable is *static* and *completely determined when it is initialized*. Taichi's compiler does type check at compile time, so you *cannot* change a variable's type once it is initialized. Still, from time to time, you may run into a situation where you have a certain type of data but it is not feasible for an assignment or calculation. Then, you need explicit type casting:

Reviewer 2: Note that statically typed is the idiomatic wording to describe a programming language.

```suggestion
In the Taichi scope, a variable is *statically typed* upon initialization. Taichi's compiler does type check at compile time, so you *cannot* change a variable's type once it is initialized. Still, from time to time, you may run into a situation where you have a certain type of data but it is not feasible for an assignment or calculation. Then, you need explicit type casting:
```

> - You can use `ti.cast()` to convert a value to the default integer or to the default floating-point type:

Reviewer: A bit inaccurate here. I believe not to default, but to any type specified in the argument.
```suggestion
- You can use `ti.cast()` to convert a value to your target type:
```

Author: I see your point now!

      c = float(b)  # 3.0
  ```

### Implicit type casting

Implicit type casting occurs when you *accidentally* put or assign a value in a place where a different data type is expected.

:::caution WARNING
As a rule of thumb, implicit type casting is a major source of bugs. And Taichi does *not* recommend resorting to this mechanism.

:::

#### Implicit type casting in binary operations

Taichi follows the [implicit conversion rules](https://en.cppreference.com/w/c/language/conversion) for the C programming language and implicitly casts operands in a [binary operation](https://en.wikipedia.org/wiki/Binary_operation) into a *common type* if the operation involves different data types. Following are two most straightforward rules for determining the common type in a binary operation:

- `i32 + f32 = f32` (`int` + `float` = `float`)
- `i32 + i64 = i64` (low precision bits + high precision bits = high precision bits)

#### Implicit type casting in assignments

When you assign a value to a variable of a different data type, Taichi implicitly casts the value into that type. Further, if the type of the variable is *not* the common type, a warning of precision loss occurs.

> - Example 1: Variable `a` is initialized with type `float` and immediately reassigned `1`. The reassignment implicitly casts `1` from `int` to `float` without warning because the data type of `a` is the common type `float`:

Reviewer 2: reassigned with `1`?

Author: Active voice: Reassign a <value> to a <variable>; Passive voice. A <variable> is reassigned a <value>.
